core/lib/src/fs/server.rs

```diff
@@ -60,351 +72,492 @@ use crate::fs::NamedFile;
+        self.rewrites.push(Arc::new(f));
+        self
+    }
 
-    /// Returns `true` if `self` is a superset of `other`. In other words,
-    /// returns `true` if all of the options in `other` are also in `self`.
+    /// Filter what files this `FileServer` will respond with
     ///
     /// # Example
     ///
-    /// ```rust
-    /// use rocket::fs::Options;
-    ///
-    /// let index_request = Options::Index | Options::DotFiles;
-    /// assert!(index_request.contains(Options::Index));
-    /// assert!(index_request.contains(Options::DotFiles));
+    /// Filter out all paths with a filename of `hidden`.
+    /// ```rust,no_run
+    /// # use rocket::{fs::FileServer, response::Redirect, uri, Rocket, Build};
+    /// # fn launch() -> Rocket<Build> {
+    /// rocket::build()
+    ///     .mount(
+    ///         "/",
+    ///         FileServer::from("static")
+    ///            .filter_file(|f| f.path.file_name() != Some("hidden".as_ref()))
+    ///     )
+    /// # }
+    /// ```
+    pub fn filter_file<F>(self, f: F) -> Self
+        where F: Fn(&File<'_, '_>) -> bool + Send + Sync + 'static
+    {
+        self.and_rewrite(FilterFile(f))
+    }
+
+    /// Transform files
     ///
-    /// let index_only = Options::Index;
-    /// assert!(index_only.contains(Options::Index));
-    /// assert!(!index_only.contains(Options::DotFiles));
+    /// # Example
     ///
-    /// let dot_only = Options::DotFiles;
-    /// assert!(dot_only.contains(Options::DotFiles));
-    /// assert!(!dot_only.contains(Options::Index));
+    /// Append `hidden` to the path of every file returned.
+    /// ```rust,no_run
+    /// # use rocket::{fs::FileServer, Build, Rocket};
+    /// # fn launch() -> Rocket<Build> {
+    /// rocket::build()
+    ///     .mount(
+    ///         "/",
+    ///         FileServer::from("static").map_file(|f| f.map_path(|p| p.join("hidden")).into())
+    ///     )
+    /// # }
     /// ```
-    #[inline]
-    pub fn contains(self, other: Options) -> bool {
-        (other.0 & self.0) == other.0
+    pub fn map_file<F>(self, f: F) -> Self
+        where F: for<'r, 'h> Fn(File<'r, 'h>) -> FileResponse<'r, 'h> + Send + Sync + 'static
+    {
+        self.and_rewrite(MapFile(f))
     }
 }
 
-/// The default set of options: `Options::Index | Options:NormalizeDirs`.
-impl Default for Options {
-    fn default() -> Self {
-        Options::Index | Options::NormalizeDirs
+impl From<FileServer> for Vec<Route> {
+    fn from(server: FileServer) -> Self {
+        // let source = figment::Source::File(server.root.clone());
+        let mut route = Route::ranked(server.rank, Method::Get, "/<path..>", server);
+        // I'd like to provide a more descriptive name, but we can't get more
+        // information out of `dyn Rewriter`
+        route.name = Some("FileServer".into());
+        vec![route]
     }
 }
 
-impl std::ops::BitOr for Options {
-    type Output = Self;
+#[crate::async_trait]
+impl Handler for FileServer {
```

Comment:
Can you model this on the implementation in my patch? I find that version easier to read.

Reply:
I've simplified the implementation a bit, and extracted some of the code to a separate function, but I don't think I can do much better. Many of the differences (the `.exists` check, and the new `strip_trailing_slash`) stem from more thorough testing.